gpu/octree/src/cuda/octree_host.cu

```diff
@@ -227,30 +278,50 @@ void  pcl::device::OctreeImpl::approxNearestSearchHost(const PointType& query, i
     float3 maxp = octreeGlobal.maxp;
 
     int node_idx = 0;
+    int code = CalcMorton(minp, maxp)(query);
+    int level = 0;
 
-    bool out_of_root = query.x < minp.x || query.y < minp.y ||  query.z < minp.z || query.x > maxp.x || query.y > maxp.y ||  query.z > maxp.z;
+    bool centroid_traversal = false;
+    int mask_pos;
+    int x, y, z;
 
-    if(!out_of_root)        
+    for(;;)
     {
-        int code = CalcMorton(minp, maxp)(query);
-        int level = 0;
+        int node = host_octree.nodes[node_idx];
+        int mask = node & 0xFF;
 
-        for(;;)
-        {            
-            int mask_pos = 1 << Morton::extractLevelCode(code, level);
+        float3 query_point;
```

Comment:
I added this because `query` is of type `PointType` in the sync gpu function (its type is `float3` in the async gpu function though). I could change this to 
```cpp
 float3 query_point = make_float3(query.x, query.y, query.z);
```

Reply:
I've checked and `PointType` is represented as `float4`. I would suggest to either: a) set `query` in `nearestVoxelTraversal` also as a `PointType`, or b) alternatively try to type-pun to be interpreted as `float3` using memcpy. With some luck, the compiler figures this out and prevents the copy. 
https://en.cppreference.com/w/cpp/string/byte/memcpy

Option b) sounds potentially prone to  "hard to track bugs" in the future